### DISCUSSION TITLE
Error "cannot find context for 'fork'"

### DISCUSSION THREAD

**Discussion by mlrennerfr** [BlueBrain/BluePyOpt#453] [Q&A]
Hi,
I'm trying to run the example codes of the jupiter notebooks but I obtain the error in the title of this post when I ask to run the cell with the line
responses = twostep_protocol.run(cell_model=simple_cell, param_values=default_params, sim=nrn)
Is it a problem of the version of Python? I'm running Anaconda with Python 3.9 and NEURON 8.0. I've already managed to run the codes without any problem some time ago in other computer (Python 2.7) so this error confuses me. 
Thank you in advance for your help!

**Comment by DrTaDa** [comment]
Hello,
It seems to be a windows related error.
I have tried to solve the issue in the branch https://github.com/BlueBrain/BluePyOpt/tree/win32_compatibility
However, I do not have a windows machine at hand to check if it works.
Could you let me know if it solves your issue ?

**Comment by mlrennerfr** [comment]
Hello,

Thank you for your quick answer! unfortunately it did not work, unless I did something wrong. I've just open the jupiter notebook for the "simple cell" from this new folder of BluePyOpt... or should I install BluePyOpt again in python?

Anyway, here is the full error message:

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_5632\804922019.py in <module>
      1 default_params = {'gnabar_hh': 0.1, 'gkbar_hh': 0.03}
----> 2 responses = twostep_protocol.run(cell_model=simple_cell, param_values=default_params, sim=nrn)

~\AppData\Roaming\Python\Python39\site-packages\bluepyopt\ephys\protocols.py in run(self, cell_model, param_values, sim, isolate, timeout)
     86             # Try/except added for backward compatibility
     87             try:
---> 88                 response = protocol.run(
     89                     cell_model=cell_model,
     90                     param_values=param_values,

~\AppData\Roaming\Python\Python39\site-packages\bluepyopt\ephys\protocols.py in inner(self, cell_model, param_values, **kwargs)
    186                     mech.deterministic = True
    187 
--> 188             responses = func(self, cell_model, param_values, **kwargs)
    189 
    190             if self.deterministic:

~\AppData\Roaming\Python\Python39\site-packages\bluepyopt\ephys\protocols.py in run(self, cell_model, param_values, sim, isolate, timeout)
    266             # Default context for python>=3.8 on macos is spawn.
    267             # Spwan context would reset NEURON properties, such as dt.
--> 268             multiprocessing_context = multiprocessing.get_context('fork')
    269 
    270             if timeout is not None:

C:\tools\Anaconda3\lib\multiprocessing\context.py in get_context(self, method)
    237             return self._actual_context
    238         else:
--> 239             return super().get_context(method)
    240 
    241     def set_start_method(self, method, force=False):

C:\tools\Anaconda3\lib\multiprocessing\context.py in get_context(self, method)
    191             ctx = _concrete_contexts[method]
    192         except KeyError:
--> 193             raise ValueError('cannot find context for %r' % method) from None
    194         ctx._check_available()
    195         return ctx

ValueError: cannot find context for 'fork'


Thank you again!

**Comment by mlrennerfr** [comment]
Solved! :)
I did not realized that I was using a computer in which Anaconda was installed for a single user! 
Installing it for all users solved this issue.

**Comment by DrTaDa** [comment]
Great !

**Comment by DrTaDa** [comment]
Did it with or without the fix ?

**Reply by MatheusHofner** [reply]
Hello there, 
I had the some problem running on win64. Reinstalling anaconda for all users did not solve. I replaced protocols.py for windows branch and I have a new error now. 
Can you help me?
Thanks 

RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: Traceback (most recent call last):
  File "C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\protocols.py", line 204, in _run_func
    cell_model.instantiate(sim=sim)
  File "C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\models.py", line 256, in instantiate
    self.instantiate_morphology(sim)
  File "C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\models.py", line 241, in instantiate_morphology
    self.morphology.instantiate(sim=sim, icell=self.icell)
  File "C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\morphologies.py", line 117, in instantiate
    raise IOError(
OSError: Morphology not found at 'simple.swc'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\ProgramData\anaconda3\lib\site-packages\pebble\common.py", line 174, in process_execute
    return function(*args, **kwargs)
  File "C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\protocols.py", line 239, in _run_func
    raise SweepProtocolException(
bluepyopt.ephys.protocols.SweepProtocolException: Failed to run Neuron Sweep Protocol


The above exception was the direct cause of the following exception:

SweepProtocolException                    Traceback (most recent call last)
Cell In[14], line 2
      1 default_params = {'gnabar_hh': 0.1, 'gkbar_hh': 0.03}
----> 2 responses = twostep_protocol.run(cell_model=simple_cell, param_values=default_params, sim=nrn)

File C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\protocols.py:89, in SequenceProtocol.run(self, cell_model, param_values, sim, isolate, timeout)
     85 for protocol in self.protocols:
     86 
     87     # Try/except added for backward compatibility
     88     try:
---> 89         response = protocol.run(
     90             cell_model=cell_model,
     91             param_values=param_values,
     92             sim=sim,
     93             isolate=isolate,
     94             timeout=timeout)
     95     except TypeError as e:
     96         if "unexpected keyword" in str(e):

File C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\protocols.py:189, in SweepProtocol.adjust_stochasticity.<locals>.inner(self, cell_model, param_values, **kwargs)
    186         previous_stoch_state.append(mech.deterministic)
    187         mech.deterministic = True
--> 189 responses = func(self, cell_model, param_values, **kwargs)
    191 if self.deterministic:
    192     for i, mech in enumerate(cell_model.mechanisms):

File C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\ephys\protocols.py:288, in SweepProtocol.run(self, cell_model, param_values, sim, isolate, timeout)
    282 tasks = pool.schedule(self._run_func, kwargs={
    283     'cell_model': cell_model,
    284     'param_values': param_values,
    285     'sim': sim},
    286     timeout=timeout)
    287 try:
--> 288     responses = tasks.result()
    289 except TimeoutError:
    290     logger.debug('SweepProtocol: task took longer than '
    291                  'timeout, will return empty response '
    292                  'for this recording')

File C:\ProgramData\anaconda3\lib\concurrent\futures\_base.py:458, in Future.result(self, timeout)
    456     raise CancelledError()
    457 elif self._state == FINISHED:
--> 458     return self.__get_result()
    459 else:
    460     raise TimeoutError()

File C:\ProgramData\anaconda3\lib\concurrent\futures\_base.py:403, in Future.__get_result(self)
    401 if self._exception:
    402     try:
--> 403         raise self._exception
    404     finally:
    405         # Break a reference cycle with the exception in self._exception
    406         self = None

SweepProtocolException: Failed to run Neuron Sweep Protocol

**Reply by DrTaDa** [reply]
Hello @MatheusHofner, 

I could not reproduce this issue as I do not have a windows machine, therefore I can't answer for certain. 

Is the simple.swc file present in the same directory as the notebook and are you executing the jupyter session from the directory in which the notebook is located ?

**Reply by MatheusHofner** [reply]
Hi,
it wasn't in the same directory. Now it's working.
Thank you very much.

**Reply by DrTaDa** [reply]
Great ! Since it works for you, I will assume that the branch [win32_compatibility](https://github.com/BlueBrain/BluePyOpt/tree/win32_compatibility) works as well.

**Reply by MatheusHofner** [reply]
I'm having this issue now, still in the same tutorial:

`FileNotFoundError                         Traceback (most recent call last)
Cell In[10], line 5
      1 import pickle
      2 # pickle.dump(results, open('results.pkl', 'w'))
      3 # results = pickle.load(open('results.pkl')
----> 5 cp = pickle.load(open('checkpoints/checkpoint.pkl'))
      6 results = (cp['population'],
      7         cp['halloffame'],
      8         cp['logbook'],
      9         cp['history'])
     12 pop, hall_of_fame, logs, hist = results

File C:\ProgramData\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:284, in _modified_open(file, *args, **kwargs)
    277 if file in {0, 1, 2}:
    278     raise ValueError(
    279         f"IPython won't let you open fd={file} by default "
    280         "as it is likely to crash IPython. If you know what you are doing, "
    281         "you can use builtins' open."
    282     )
--> 284 return io_open(file, *args, **kwargs)

FileNotFoundError: [Errno 2] No such file or directory: 'checkpoints/checkpoint.pkl'`

**Reply by DrTaDa** [reply]
If you wish to generate a pickle checkpoint file for the run, you have to provide a filepath for it, for example, in the jupyter notebook, you can replace the line
```
final_pop, hall_of_fame, logs, hist = optimisation.run(max_ngen=5)
```

by
```
final_pop, hall_of_fame, logs, hist = optimisation.run(max_ngen=5, cp_filename="checkpoints/checkpoint.pkl")
```

**Reply by MatheusHofner** [reply]
So the line on tutorial that creates the filepath was comment. I uncomment that, and now I got: 

FileNotFoundError                         Traceback (most recent call last)
Cell In[15], line 6
      1 optimisation = bpop.optimisations.DEAPOptimisation(
      2     evaluator=cell_evaluator,
      3     offspring_size = 100,
      4     seed=1)
----> 6 results = optimisation.run(max_ngen=10, cp_filename='checkpoints/checkpoint.pkl')

File C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\deapext\optimisations.py:309, in DEAPOptimisation.run(self, max_ngen, offspring_size, continue_cp, cp_filename, cp_frequency, cp_period, parent_population, terminator)
    306 if hasattr(self.evaluator, "param_names"):
    307     param_names = self.evaluator.param_names
--> 309 pop, hof, log, history = algorithms.eaAlphaMuPlusLambdaCheckpoint(
    310     pop,
    311     self.toolbox,
    312     offspring_size,
    313     self.cxpb,
    314     self.mutpb,
    315     max_ngen,
    316     stats=stats,
    317     halloffame=self.hof,
    318     cp_frequency=cp_frequency,
    319     cp_period=None,
    320     continue_cp=continue_cp,
    321     cp_filename=cp_filename,
    322     terminator=terminator,
    323     param_names=param_names)
    325 # Update hall of fame
    326 self.hof = hof

File C:\ProgramData\anaconda3\lib\site-packages\bluepyopt\deapext\algorithms.py:194, in eaAlphaMuPlusLambdaCheckpoint(population, toolbox, mu, cxpb, mutpb, ngen, stats, halloffame, cp_frequency, cp_period, cp_filename, continue_cp, terminator, param_names)
    183 if (cp_filename and cp_frequency and
    184    gen % cp_frequency == 0 and
    185    (cp_period is None or time.time() - time_last_save > cp_period)):
    186     cp = dict(population=population,
    187               generation=gen,
    188               parents=parents,
   (...)
    192               rndstate=random.getstate(),
    193               param_names=param_names)
--> 194     pickle.dump(cp, open(cp_filename_tmp, "wb"))
    195     if os.path.isfile(cp_filename_tmp):
    196         shutil.copy(cp_filename_tmp, cp_filename)

FileNotFoundError: [Errno 2] No such file or directory: 'checkpoints/checkpoint.pkl.tmp'

**Reply by DrTaDa** [reply]
Was the file checkpoints/checkpoint.pkl.tmp generated ?
If not, it can be a permission issue.
